lib/utils.js

```diff
@@ -142,84 +141,77 @@ function getModule(filename, base) {
   if (!stack || !Array.isArray(stack) || !stack.length || !stack[0].getFileName) {
-    // lol, stack is fucked
-    return cb(frames);
+    // the stack is not the useful thing we were expecting :/
+    return cb([]);
   }
 
-  var callbacks = stack.length;
-
-  // Sentry requires the stack trace to be from oldest to newest
+  // Sentry expects the stack trace to be oldest -> newest, v8 provides newest -> oldest
   stack.reverse();
 
-  return stack.forEach(function (line, index) {
+  var frames = [];
+  var filesToRead = {};
+  stack.forEach(function (line) {
     var frame = {
-          filename: line.getFileName() || '',
-          lineno: line.getLineNumber(),
-          colno: line.getColumnNumber(),
-          'function': getFunction(line),
-        },
-        isInternal = line.isNative() ||
-          frame.filename[0] !== '/' &&
-          frame.filename[0] !== '.' &&
-          frame.filename.indexOf(':\\') !== 1;
+      filename: line.getFileName() || '',
+      lineno: line.getLineNumber(),
+      colno: line.getColumnNumber(),
+      'function': getFunction(line),
+    };
+
+    var isInternal = line.isNative() ||
+      frame.filename[0] !== '/' &&
+      frame.filename[0] !== '.' &&
+      frame.filename.indexOf(':\\') !== 1;
 
     // in_app is all that's not an internal Node function or a module within node_modules
     // note that isNative appears to return true even for node core libraries
     // see https://github.com/getsentry/raven-node/issues/176
     frame.in_app = !isInternal && frame.filename.indexOf('node_modules/') === -1;
 
     // Extract a module name based on the filename
-    if (frame.filename) frame.module = getModule(frame.filename);
-
-    // internal Node files are not full path names. Ignore them.
-    if (isInternal) {
-      frames[index] = frame;
-      if (--callbacks === 0) cb(frames);
-      return;
+    if (frame.filename) {
+      frame.module = getModule(frame.filename);
+      if (!isInternal) filesToRead[frame.filename] = true;
     }
 
-    if (!cache[frame.filename]) {
-      cache[frame.filename] = readFilePromise(frame.filename)
-        .then(function (file) {
-          return file.toString().split('\n');
-        }, function () {
-          return null;
-        });
-    }
+    frames.push(frame);
+  });
 
-    cache[frame.filename].then(function (file) {
-      if (file) {
-        parseLines(file, frame);
+  return readSourceFiles(Object.keys(filesToRead), function (sourceFiles) {
+    frames.forEach(function (frame) {
+      if (frame.filename && sourceFiles[frame.filename]) {
+        var lines = sourceFiles[frame.filename];
+        try {
+          frame.pre_context = lines.slice(Math.max(0, frame.lineno - (LINES_OF_CONTEXT + 1)), frame.lineno - 1);
+          frame.context_line = lines[frame.lineno - 1];
+          frame.post_context = lines.slice(frame.lineno, frame.lineno + LINES_OF_CONTEXT);
+        } catch (e) {
+          // anomaly, being defensive in case
```

Comment:
nah, i think we'd rather not send them to the server at all; not sure what behavior diff would be but not sending them at all is what we do for native frames so i'll stick to that